transform_spans.go

```diff
@@ -122,6 +132,108 @@ func ocAttributesToProtoAttributes(attrs map[string]interface{}) *tracepb.Span_A
 	}
 }
 
+// the following code are mostly from oc stackdriver
+// see https://github.com/census-ecosystem/opencensus-go-exporter-stackdriver/blob/master/trace_proto.go#L46 for detail
+func ocTimeEventsToProtoTimeEvents(as []trace.Annotation, es []trace.MessageEvent) *tracepb.Span_TimeEvents {
+	if len(as) == 0 && len(es) == 0 {
+		return nil
+	}
+
+	timeEvents := &tracepb.Span_TimeEvents{}
+	var annotations, droppedAnnotationsCount int
+	var messageEvents, droppedMessageEventsCount int
+
+	// transform annotations
+	for i, a := range as {
+		if annotations >= maxAnnotationEventsPerSpan {
+			droppedAnnotationsCount = len(as) - i
+			break
+		}
+		annotations++
+		timeEvents.TimeEvent = append(timeEvents.TimeEvent,
+			&tracepb.Span_TimeEvent{
+				Time:  timeToTimestamp(a.Time),
+				Value: transformAnnoationToTimeEvent(&a),
+			},
+		)
+	}
+
+	// transform message events
+	for i, e := range es {
+		if messageEvents >= maxMessageEventsPerSpan {
+			droppedMessageEventsCount = len(es) - i
+			break
+		}
+		messageEvents++
+		timeEvents.TimeEvent = append(timeEvents.TimeEvent,
+			&tracepb.Span_TimeEvent{
+				Time:  timeToTimestamp(e.Time),
+				Value: transformMessageEventToTimeEvent(&e),
+			},
+		)
+	}
+
+	// process dropped counter
+	timeEvents.DroppedAnnotationsCount = clip32(droppedAnnotationsCount)
+	timeEvents.DroppedMessageEventsCount = clip32(droppedMessageEventsCount)
+
+	return timeEvents
+}
+
+func transformAnnoationToTimeEvent(a *trace.Annotation) *tracepb.Span_TimeEvent_Annotation_ {
+	return &tracepb.Span_TimeEvent_Annotation_{
+		Annotation: &tracepb.Span_TimeEvent_Annotation{
+			Description: trunc(a.Message, maxAttributeStringValue),
+			Attributes:  ocAttributesToProtoAttributes(a.Attributes),
+		},
+	}
+}
+
+func transformMessageEventToTimeEvent(e *trace.MessageEvent) *tracepb.Span_TimeEvent_MessageEvent_ {
+	return &tracepb.Span_TimeEvent_MessageEvent_{
+		MessageEvent: &tracepb.Span_TimeEvent_MessageEvent{
+			Type:             tracepb.Span_TimeEvent_MessageEvent_Type(e.EventType),
+			Id:               uint64(e.MessageID),
+			UncompressedSize: uint64(e.UncompressedByteSize),
+			CompressedSize:   uint64(e.CompressedByteSize),
+		},
+	}
+}
+
+// trunc returns a TruncatableString truncated to the given limit.
+func trunc(s string, limit int) *tracepb.TruncatableString {
+	if len(s) > limit {
+		b := []byte(s[:limit])
+		for {
+			r, size := utf8.DecodeLastRune(b)
+			if r == utf8.RuneError && size == 1 {
+				b = b[:len(b)-1]
+			} else {
+				break
```

Comment:
So what did you intend to do in this loop?
This just looks like it checks only the last rune once and then breaks if only the last rune is a valid UTF-8 rune and only truncates off invalid UTF-8 runes but from the terminal end. The code would serve the purpose if perhaps you kept track of the characters that you've seen before but as is, it'll only run once for anything with a valid last rune.

To truncate a UTF-8 string, I believe we have to start from the front, keep a counter of what we've seen before but also truncate to a max of `limit` so here is a suggestion
```go
// truncateUTF8 walks a presumable valid UTF-8 sequence from
// the left and truncates it to an upperbound of max whereby
// if a rune's size + length of already accepted string > max
// then that rune won't be included.
func truncateUTF8(b []byte, max int) []byte {
	if len(b) <= max {
		return b
	}

	// Otherwise performing the truncation

	i := 0
	for i < len(b) {
		r, size := utf8.DecodeRune(b[i:])
		if size == 0 { // Empty sequence
			break
		}
		if r == utf8.RuneError {
			break
		}
		if i+size > max {
			break
		}
		i += size
	}
	return b[:i]
}
```
and here is a working sample https://play.golang.org/p/dcpjtvXBdOo

Reply:
Please ignore that change request as I noticed in https://github.com/census-ecosystem/opencensus-go-exporter-ocagent/pull/30#discussion_r227565941 that the spec doesn't have a limit to the name hence please remove these checks entirely.

Reply:
I thought it would be faster if DecodeRune from end to start for large truncate length, but when i made some benchmark, there are no significant difference. So please ignore that:)

Here are the test code: https://play.golang.org/p/8ahSADIxeB6